### PR TITLE
Tensor changes

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add C++17 implementation of std::array. Can be removed once we have C++17
- Move from TMP to constexpr metaprogramming in Tensor where possible to allow for higher rank and higher dimensional Tensor's. This is required for #495 

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: src/Utilities/ConstantExpressions.hpp [hunk 1]
```diff
@@ -53,6 +53,25 @@ SPECTRE_ALWAYS_INLINE constexpr decltype(auto) cube(const T& x) {
...
 }
 
+/*!
+ * \ingroup ConstantExpressionsGroup
+ * \brief Compute the factorial, or ratios of factorials
+ *
+ * \note The largest representable factorial is 20!. For ratios it is up to the
+ * user to ensure that overflow does not occur.
+ */
+SPECTRE_ALWAYS_INLINE constexpr size_t factorial(
```
Comment (kidder) [inline src/Utilities/ConstantExpressions.hpp]: falling_factorial with x and n linked to MathWhatever

File: src/Utilities/ConstantExpressions.hpp [hunk 2]
```diff
@@ -53,6 +53,25 @@ SPECTRE_ALWAYS_INLINE constexpr decltype(auto) cube(const T& x) {
...
+    const size_t t, const size_t divided_out = 1) noexcept {
+  // clang-tidy: don't warn about STL internals, I can't fix them
+  assert(divided_out > 0);                    // NOLINT
```
Comment (kidder) [inline src/Utilities/ConstantExpressions.hpp]: assert n<=x

File: tests/Unit/Utilities/Test_ConstantExpressions.cpp [hunk 3]
```diff
@@ -50,6 +50,14 @@ static_assert(get_nth_bit(2, 1) == 1,
 static_assert(get_nth_bit(4, 2) == 1,
               "Failed test Unit.Utilities.ConstantExpressions");
 
+// Test factorial
+static_assert(factorial(3) == 6, "");
```
Comment (kidder) [inline tests/Unit/Utilities/Test_ConstantExpressions.cpp]: messages

File: src/DataStructures/Tensor/Structure.hpp [hunk 4]
```diff
@@ -38,29 +38,39 @@ template <size_t Size>
...
+      if (static_cast<size_t>(symm[i]) == symm_index) {  // NOLINT
+        ++number_of_indices_with_symm;
+        dim_of_index = dims[i];  // NOLINT
+      }
     }
+    assert(dim_of_index > 0);                 // NOLINT
+    assert(number_of_indices_with_symm > 0);  // NOLINT
+    total_independent_components *=
+        factorial(dim_of_index + number_of_indices_with_symm - 1) /
+        (factorial(dim_of_index - 1) * factorial(number_of_indices_with_symm));
```
Comment (kidder) [inline src/DataStructures/Tensor/Structure.hpp]: use falling_factorial

File: tests/Unit/DataStructures/Test_Tensor.cpp [hunk 5]
```diff
@@ -515,6 +512,402 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.RankAndSize",
       symmetric_rank3_dim4{};
   CHECK(3 == symmetric_rank3_dim4.rank());
   CHECK(40 == symmetric_rank3_dim4.size());
+  i = 1.948;  // not a value likely to be default-constructed to
+  for (auto& p : symmetric_rank3_dim4) {
+    p = i;
+    ++i;
+  }
+  CHECK(symmetric_rank3_dim4.get(0, 0, 1) == symmetric_rank3_dim4.get(0, 1, 0));
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_Tensor.cpp]: loop over all first indices

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
@wthrowe can you look at this

File: src/Utilities/ConstantExpressions.hpp (hunk 1 above)
Comment (wthrowe) [reply src/Utilities/ConstantExpressions.hpp]: I find the argument names completely opaque.  (I think this is what the comment above is saying, but I'm not sure.)

File: src/DataStructures/Tensor/Structure.hpp [hunk 6]
```diff
@@ -38,29 +38,39 @@ template <size_t Size>
...
-    if (component_number == static_cast<size_t>(ce_abs(first_symm))) {
-      component_number++;
-    } else {
-      number /= 2;
+  size_t max_element = 0;
+  for (size_t i = 0; i < Size; ++i) {
+    assert(symm[i] > 0);  // NOLINT
+    max_element = static_cast<size_t>(symm[i]) > max_element  // NOLINT
+                      ? static_cast<size_t>(symm[i])          // NOLINT
+                      : max_element;
```
Comment (wthrowe) [inline src/DataStructures/Tensor/Structure.hpp]: `std::max(...)`

File: tests/Unit/DataStructures/Test_Tensor.cpp [hunk 7]
```diff
@@ -502,6 +488,17 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.RankAndSize",
...
+  for (auto& p : symmetric_rank2_dim4) {
+    p = i;
+    ++i;
+  }
+  CHECK(symmetric_rank2_dim4.get(0, 1) == symmetric_rank2_dim4.get(1, 0));
+  CHECK(symmetric_rank2_dim4.get(0, 2) == symmetric_rank2_dim4.get(2, 0));
+  CHECK(symmetric_rank2_dim4.get(0, 3) == symmetric_rank2_dim4.get(3, 0));
+  CHECK(symmetric_rank2_dim4.get(2, 1) == symmetric_rank2_dim4.get(1, 2));
+  CHECK(symmetric_rank2_dim4.get(3, 1) == symmetric_rank2_dim4.get(1, 3));
+  CHECK(symmetric_rank2_dim4.get(2, 3) == symmetric_rank2_dim4.get(3, 2));
```
Comment (wthrowe) [inline tests/Unit/DataStructures/Test_Tensor.cpp]: Add a check somewhere that some pair of entries that should be different are different.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/DataStructures/Tensor/Structure.hpp [hunk 8]
```diff
@@ -40,10 +40,11 @@ constexpr size_t number_of_independent_components(
     const std::array<size_t, Size>& dims) noexcept {
   size_t max_element = 0;
   for (size_t i = 0; i < Size; ++i) {
+    // clang-tidy: internal STL and gls::at (don't need at in constexpr)
```
Comment (kidder) [inline src/DataStructures/Tensor/Structure.hpp]: need at -> need it

File: src/DataStructures/Tensor/Structure.hpp [hunk 9]
```diff
@@ -40,10 +40,11 @@ constexpr size_t number_of_independent_components(
...
     assert(symm[i] > 0);  // NOLINT
-    max_element = static_cast<size_t>(symm[i]) > max_element  // NOLINT
-                      ? static_cast<size_t>(symm[i])          // NOLINT
-                      : max_element;
+    max_element = std::max(
+        static_cast<size_t>(symm[i] < 0 ? -symm[i] : symm[i]),  // NOLINT
```
Comment (kidder) [inline src/DataStructures/Tensor/Structure.hpp]: use abs

**Review by kidder — CHANGES_REQUESTED** [review]
squash

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**